File: cyclops/connectors/abuseipdb.py
import os
import aiohttp
from cyclops.connectors.base import BaseConnector


class AbuseIPDBConnector(BaseConnector):
    def __init__(self, config, state, telemetry):
        super().__init__(config, state, telemetry)
        self.name = "AbuseIPDB"
        self.api_key = os.environ.get("ABUSEIPDB_API_KEY")

    async def run(self, investigation_id, phase_id, input_type, input_value):
        if not self.api_key:
            return {"status": "skipped", "error": "ABUSEIPDB_API_KEY not set"}
        run_id = self.state.record_connector_run(investigation_id, phase_id, self.name, input_type, input_value)
        self.telemetry.connector_start(investigation_id, self.name, phase_id, {"type": input_type, "value": input_value})
        try:
            entities = []
            async with aiohttp.ClientSession(headers={"Key": self.api_key, "Accept": "application/json"}) as session:
                async with session.get(f"https://api.abuseipdb.com/api/v2/check?ipAddress={input_value}&maxAgeInDays=90&verbose", timeout=aiohttp.ClientTimeout(total=10)) as resp:
                    if resp.status >= 400:
                        raise Exception(f"AbuseIPDB HTTP {resp.status}")
                    data = await resp.json()
            d = data.get("data", {})
            if d:
                entities.append({"type": "ip", "data": {"address": input_value, "abuse_score": d.get("abuseConfidenceScore"), "country": d.get("countryCode"),
                    "isp": d.get("isp"), "domain": d.get("domain"), "usage_type": d.get("usageType"), "is_tor": d.get("isTor"),
                    "total_reports": d.get("totalReports"), "last_reported": d.get("lastReportedAt"), "source": "abuseipdb"}, "confidence": 0.9})
                if d.get("domain"):
                    entities.append({"type": "domain", "data": {"name": d["domain"], "source": "abuseipdb"}, "confidence": 0.6, "as_known": {"type": "domain", "value": d["domain"]}})
            new_count = 0
            for entity in entities:
                added = self.state.add_entity(investigation_id, entity["type"], entity["data"], entity.get("confidence", 0.5))
                if added["new"]:
                    new_count += 1
                self.telemetry.entity_discovered(investigation_id, entity["type"], added["new"], self.name)
                if "as_known" in entity:
                    self.state.add_known(investigation_id, entity["as_known"]["type"], entity["as_known"]["value"], self.name)
            self.state.complete_connector_run(run_id, "completed", new_count)
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "completed", "entitiesFound": new_count, "input": {"type": input_type, "value": input_value}})
            return {"status": "completed", "entities": entities, "newCount": new_count}
        except Exception as e:
            self.state.complete_connector_run(run_id, "failed", 0, error=str(e))
            self.telemetry.connector_end(investigation_id, self.name, phase_id, {"status": "failed", "input": {"type": input_type, "value": input_value}})
            return {"status": "failed", "error": str(e)}

    async def health_check(self):
        return {"ok": bool(self.api_key), "hasApiKey": bool(self.api_key)}

File: src/connectors/abuseipdb.js
import { BaseConnector } from './base.js';
import https from 'https';

export class AbuseIPDBConnector extends BaseConnector {
    constructor(config, state, telemetry) {
        super(config, state, telemetry);
        this.name = 'AbuseIPDB';
        this.apiKey = process.env.ABUSEIPDB_API_KEY;
    }

    async run(investigationId, phaseId, inputType, inputValue) {
        if (!this.apiKey) return { status: 'skipped', error: 'ABUSEIPDB_API_KEY not set' };

        const runId = this.state.recordConnectorRun(investigationId, phaseId, this.name, inputType, inputValue);
        this.telemetry.connectorStart(investigationId, this.name, phaseId, { type: inputType, value: inputValue });

        try {
            const entities = [];
            const data = await this._apiGet(`/api/v2/check?ipAddress=${encodeURIComponent(inputValue)}&maxAgeInDays=90&verbose`);

            if (data?.data) {
                const d = data.data;
                entities.push({
                    type: 'ip',
                    data: {
                        address: inputValue,
                        abuse_score: d.abuseConfidenceScore,
                        country: d.countryCode,
                        isp: d.isp,
                        domain: d.domain,
                        usage_type: d.usageType,
                        is_tor: d.isTor,
                        is_proxy: d.isProxy || false,
                        total_reports: d.totalReports,
                        num_distinct_users: d.numDistinctUsers,
                        last_reported: d.lastReportedAt,
                        source: 'abuseipdb'
                    },
                    confidence: 0.9
                });

                if (d.domain) {
                    entities.push({
                        type: 'domain',
                        data: { name: d.domain, source: 'abuseipdb_isp' },
                        confidence: 0.6,
                        asKnown: { type: 'domain', value: d.domain }
                    });
                }

                if (d.hostnames?.length) {
                    for (const hostname of d.hostnames) {
                        entities.push({
                            type: 'domain',
                            data: { name: hostname, source: 'abuseipdb_hostname' },
                            confidence: 0.8,
                            asKnown: { type: 'domain', value: hostname }
                        });
                    }
                }
            }

            let newCount = 0;
            for (const entity of entities) {
                const added = this.state.addEntity(investigationId, entity.type, entity.data, entity.confidence);
                if (added.new) newCount++;
                this.telemetry.entityDiscovered(investigationId, entity.type, added.new, this.name);
                if (entity.asKnown) {
                    this.state.addKnown(investigationId, entity.asKnown.type, entity.asKnown.value, this.name, entity.confidence);
                }
            }

            this.state.completeConnectorRun(runId, 'completed', newCount, null, null, 0);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'completed', entitiesFound: newCount, input: { type: inputType, value: inputValue }
            });
            return { status: 'completed', entities, newCount };
        } catch (e) {
            this.state.completeConnectorRun(runId, 'failed', 0, null, e.message, 1);
            this.telemetry.connectorEnd(investigationId, this.name, phaseId, {
                status: 'failed', entitiesFound: 0, input: { type: inputType, value: inputValue }
            });
            return { status: 'failed', error: e.message };
        }
    }

    _apiGet(path) {
        return new Promise((resolve, reject) => {
            const req = https.get({
                hostname: 'api.abuseipdb.com',
                path,
                headers: { 'Key': this.apiKey, 'Accept': 'application/json' },
                timeout: 10000
            }, res => {
                if (res.statusCode === 404) { res.resume(); return resolve(null); }
                if (res.statusCode === 429) { res.resume(); return reject(new Error('AbuseIPDB rate limited')); }
                if (res.statusCode >= 400) { res.resume(); return reject(new Error(`AbuseIPDB HTTP ${res.statusCode}`)); }
                let data = '';
                res.on('data', chunk => { data += chunk; });
                res.on('end', () => {
                    try { resolve(JSON.parse(data)); } catch { resolve(null); }
                });
            });
            req.on('error', reject);
            req.on('timeout', () => { req.destroy(); reject(new Error('AbuseIPDB timeout')); });
        });
    }

    async healthCheck() {
        return { ok: !!this.apiKey, version: 'v2', hasApiKey: !!this.apiKey };
    }
}
